File: Codewars/Valid Braces.cpp
// Valid Braces
// https://www.codewars.com/kata/valid-braces/
// by Equim on 16-12-14

#include <map>
#include <stack>

using namespace std;
bool valid_braces(string braces)
{
    map<char, char> KDict;
    KDict['('] = ')';
    KDict['['] = ']';
    KDict['{'] = '}';
    
    stack<char> K;

    for (int i = 0;i < braces.length();i++)
    {
        if (braces[i] == '(' || braces[i] == '[' || braces[i] == '{')
            K.push(braces[i]);
        else if (K.empty())
            return false;
        else
        {
            char current = K.top();
            K.pop();
            if (KDict[current] != braces[i])
                return false;
        }
    }
    return K.empty();
}

File: Codewars/Valid Braces.js
// Valid Braces
// https://www.codewars.com/kata/valid-braces/
// by Equim on 16-12-14

function validBraces(braces)
{
    var KDict = { '(': ')', '[': ']', '{': '}' };
    var K = [];

    for (var i = 0;i < braces.length;i++)
    {
        if (braces.charAt(i) == '(' || braces.charAt(i) == '[' || braces.charAt(i) == '{')
            K.push(braces.charAt(i));
        else if (K.length == 0 || KDict[K.pop()] != braces.charAt(i))
            return false;
    }
    return K.length == 0;
}
